### DISCUSSION TITLE
Allow `gatsby-*.js` of a plugin to be placed in a custom directory

### DISCUSSION THREAD

**Discussion by alvis** [gatsbyjs/gatsby#32132] [Ideas / Feature Requests]
Currently, gatsby requires all `gatsby-*.js` files presents on the root directory of a plugin.

While it may be fine for pure js, it violates the best practice if the source of the plugin requires compilation such as those written in typescript. Most of these plugin may have their compiled `gatsby-*.js` files stored in other directories such as `lib`. As such, many such plugins have to place a dummy `gatsby-*.js` files on the root with content like:
```ts
module.exports = require('./lib/gatsby-node')
```

Since ususally in `package.json` of these plugin contains the output directory in `main` (usually `lib/index.js`), can we get gatsby to look for `gatsby-*.js` in the same directory of `main` if they are not present in the root directory?

**Comment by wardpeet** [comment]
Hi, we're planning to introduce this feature soon. I can't give you an exact date yet but we will be leveraging the exports field in package.json
https://nodejs.org/api/packages.html#packages_subpath_patterns

As an example by setting the following fields in your package.json
```json
"name": "something",
"version": "1.0.0",
"exports": {
  "./gatsby-node.js": "./lib/gatsby-node.js"
}
```
